Dockerfile

```diff
@@ -1,11 +1,13 @@
 FROM haskell:8.10.4 as base
 
 RUN apt-get update && apt-get install -y \
+    graphviz \
     libblas-dev \
     libgsl-dev \
     liblapack-dev \
     libssl-dev \
-    openssl
+    openssl \
+    python3
 
 RUN cabal update
```

Comment:
I'm not a fan of putting a bare call to `cabal update` in a Dockerfile, as it will cause different packages to be used depending on when the Dockerfile is built. I'd prefer passing a date to `cabal update` to fix the package index it uses to avoid this problem. Something like `cabal update 'hackage.haskell.org,2020-05-09T21:53:25Z'`, but with a more recent date.

On that note, the call to `cabal update ...` should be combined with the call to `cabal v2-build` below (i.e., `cabal v2-update <date> && cabal v2-build ...`), similarly to how we combine the calls to `apt-get update` and `apt-get install`.

Reply:
Thanks, I didn't know that was an ability of `cabal update`! I'm a little curious about the rationale for making both that change and combining the `update` and `install` calls, though. It seems like the reasoning behind the `<package-manager> update && <package-manager> install` idiom is that `<package-manager> update`, despite appearing to Docker to be an idempotent operation, isn't - but as far as I can tell, `cabal update <date>` is. Is there another facet of the idiom that would warrant its application here?

Reply:
A good point, I suppose `update && install` only makes sense if `update` isn't idempotent. In that case, feel free to disregard that suggestion.

Reply:
Cool. I'm mostly minded to do that since it allows keeping `cabal update` above `COPY`ing the cabal file, which would allow a minor build time improvement and let more builds succeed without network connectivity.